### PR TITLE
Icinga DB: introduce icinga:*:state#normalized_performance_data

### PR DESCRIPTION
**Pull request by Al2Klimov**

### CODE REVIEW

**Comment by Al2Klimov** [comment]
## Blocked by

* [ ] #8759

**Comment by Al2Klimov** [comment]
Does DB web care for the non-normalized perfdata? If no: does it make sense to use the already existing column?

**Comment by Al2Klimov** [comment]
We _might_... OK... but we don’t yet. Shall I reduce the doc changes here to just that IDB gets the normalized ones?

**Comment by Al2Klimov** [comment]
## Test

```
$ redis-cli
127.0.0.1:6379> HGETALL icinga:host:state
1) "5ffd3470e558666b25ccd505c329faa0f1932596"
2) "{\"acknowledgement\":0,\"check_attempt\":1,\"check_source\":\"Alexanders-MacBook-Pro.local\",\"check_timeout\":60000,\"commandline\":\"'dummy'\",\"environment_id\":\"da39a3ee5e6b4b0d3255bfef95601890afd80709\",\"execution_time\":0,\"hard_state\":0,\"id\":\"5ffd3470e558666b25ccd505c329faa0f1932596\",\"in_downtime\":false,\"is_active\":true,\"is_flapping\":false,\"is_handled\":false,\"is_problem\":false,\"is_reachable\":true,\"last_state_change\":1620744711202,\"last_update\":1620744711202,\"latency\":6,\"next_check\":1620744769972,\"next_update\":1620744829985,\"normalized_performance_data\":\"lolcat=60s\",\"object_id\":\"5ffd3470e558666b25ccd505c329faa0f1932596\",\"output\":\" \",\"performance_data\":\"lolcat=1m\",\"previous_hard_state\":99,\"severity\":0,\"state\":0,\"state_type\":1}"
3) "d05af40f3175e2dd51f25a555506752280d68bd3"
4) "{\"acknowledgement\":0,\"check_attempt\":1,\"check_source\":\"Alexanders-MacBook-Pro.local\",\"check_timeout\":60000,\"commandline\":\"'echo \\\\' |lolcat=2m\\\\''\",\"environment_id\":\"da39a3ee5e6b4b0d3255bfef95601890afd80709\",\"execution_time\":10,\"hard_state\":0,\"id\":\"d05af40f3175e2dd51f25a555506752280d68bd3\",\"in_downtime\":false,\"is_active\":true,\"is_flapping\":false,\"is_handled\":false,\"is_problem\":false,\"is_reachable\":true,\"last_state_change\":1620744703326,\"last_update\":1620744703326,\"latency\":12,\"next_check\":1620744990096,\"next_update\":1620745290141,\"normalized_performance_data\":\"lolcat=120s\",\"object_id\":\"d05af40f3175e2dd51f25a555506752280d68bd3\",\"performance_data\":\"lolcat=2m\",\"previous_hard_state\":99,\"severity\":0,\"state\":0,\"state_type\":1}"
127.0.0.1:6379>
```

Purrs like a cat:

```
\"normalized_performance_data\":\"lolcat=60s\"
```

```
\"performance_data\":\"lolcat=1m\"
```

**Comment by julianbrost** [comment]
Can you please rebase this so that GitHub doesn't show 3bc9969acb39197a0c0a3101d7b511ee87750fc7/#8759 as part of this PR?

File: lib/icinga/pluginutility.cpp [hunk 1]
```diff
@@ -192,10 +192,25 @@ String PluginUtility::FormatPerfdata(const Array::Ptr& perfdata)
...
+				normalized = PerfdataValue::Parse(pdv);
+			} catch (const std::invalid_argument& ex) {
+				Log(LogDebug, "PerfdataValue") << ex.what();
+			}
+
+			if (normalized) {
+				result << normalized->Format();
+			} else {
+				result << pdv;
+			}
```
Comment (julianbrost) [inline lib/icinga/pluginutility.cpp:210]: Can't this be just done within the try/catch? Shouldn't make a difference unless `normalized->Format()` is supposed to throw `std::invalid_argument` and this should be passed to the caller.

File: test/icinga-perfdata.cpp [hunk 2]
```diff
@@ -43,6 +43,15 @@ BOOST_AUTO_TEST_CASE(multiple)
 	BOOST_CHECK(str == "testA=123456 testB=123456");
 }
 
+BOOST_AUTO_TEST_CASE(normalize)
```
Comment (julianbrost) [inline test/icinga-perfdata.cpp:46]: Has to be added to `CMakeLists.txt`

File: test/icinga-perfdata.cpp [hunk 3]
```diff
@@ -43,6 +43,15 @@ BOOST_AUTO_TEST_CASE(multiple)
...
+{
+	Array::Ptr pd = PluginUtility::SplitPerfdata("testA=2m;3;4;1;5 testB=2foobar testC");
```
Comment (julianbrost) [inline test/icinga-perfdata.cpp:48]: What's `testC` supposed to test? Doesn't look like valid perfdata to me.

File: doc/05-service-monitoring.md [hunk 4]
```diff
@@ -651,7 +651,7 @@ Icinga sets `LC_NUMERIC=C` to enforce this locale on plugin execution.
 ```
 
 The UoMs are written as-is into the [core backends](14-features.md#core-backends)
-(Icinga DB, IDO, API). I.e. 12.445000ms remain 12.445000ms.
+(IDO, API). I.e. 12.445000ms remain 12.445000ms.
```
Comment (julianbrost) [inline doc/05-service-monitoring.md]: Now no longer affects all core backends, given that Icinga DB is out. Also, would it make sense to expose the normalized perfdata over the API?

**Review by julianbrost — CHANGES_REQUESTED** [review]
Since you have to update the PR now anyways, please rebase so that 3bc9969acb39197a0c0a3101d7b511ee87750fc7 is no longer considered to be part of this PR.

File: lib/icinga/pluginutility.cpp (hunk 1 above)
Comment (Al2Klimov) [reply lib/icinga/pluginutility.cpp:210]: I, however, prefer as less as possible in a try/catch.

File: doc/05-service-monitoring.md (hunk 4 above)
Comment (Al2Klimov) [reply doc/05-service-monitoring.md:654]: Once someone needs it...

**Review by julianbrost — APPROVED** [review]

**Pull request merged**